client/peerset/src/lib.rs

```diff
@@ -615,165 +327,106 @@ impl Peerset {
+			if let Some(action) = action {
+				match action {
+					Action::AddReservedPeer(set_id, peer_id) =>
+						self.protocol_handles[set_id.0].add_reserved_peer(peer_id),
+					Action::RemoveReservedPeer(set_id, peer_id) =>
+						self.protocol_handles[set_id.0].remove_reserved_peer(peer_id),
+					Action::SetReservedPeers(set_id, peer_ids) =>
+						self.protocol_handles[set_id.0].set_reserved_peers(peer_ids),
+					Action::SetReservedOnly(set_id, reserved_only) =>
+						self.protocol_handles[set_id.0].set_reserved_only(reserved_only),
+					Action::ReportPeer(peer_id, score_diff) =>
+						self.peer_store_handle.report_peer(peer_id, score_diff),
+					Action::AddKnownPeer(peer_id) => self.peer_store_handle.add_known_peer(peer_id),
+					Action::PeerReputation(peer_id, pending_response) => {
+						let _ =
+							pending_response.send(self.peer_store_handle.peer_reputation(&peer_id));
+					},
 				}
+			} else {
+				debug!(target: LOG_TARGET, "`PeersetHandle` was dropped, terminating `Peerset`.");
+				return Poll::Ready(None)
 			}
+		}
 
-			let action = match Stream::poll_next(Pin::new(&mut self.rx), cx) {
-				Poll::Pending => return Poll::Pending,
-				Poll::Ready(Some(event)) => event,
-				Poll::Ready(None) => return Poll::Pending,
-			};
-
-			match action {
-				Action::AddReservedPeer(set_id, peer_id) =>
-					self.on_add_reserved_peer(set_id, peer_id),
-				Action::RemoveReservedPeer(set_id, peer_id) =>
-					self.on_remove_reserved_peer(set_id, peer_id),
-				Action::SetReservedPeers(set_id, peer_ids) =>
-					self.on_set_reserved_peers(set_id, peer_ids),
-				Action::SetReservedOnly(set_id, reserved) =>
-					self.on_set_reserved_only(set_id, reserved),
-				Action::ReportPeer(peer_id, score_diff) => self.on_report_peer(peer_id, score_diff),
-				Action::AddToPeersSet(sets_name, peer_id) =>
-					self.add_to_peers_set(sets_name, peer_id),
-				Action::RemoveFromPeersSet(sets_name, peer_id) =>
-					self.on_remove_from_peers_set(sets_name, peer_id),
-				Action::PeerReputation(peer_id, pending_response) =>
-					self.on_peer_reputation(peer_id, pending_response),
-			}
+		if let Poll::Ready(()) = self.peer_store_future.poll_unpin(cx) {
+			debug!(target: LOG_TARGET, "`PeerStore` has terminated, terminating `PeerSet`.");
+			return Poll::Ready(None)
+		}
+
+		if let Poll::Ready(_) = self.protocol_controller_futures.poll_unpin(cx) {
+			debug!(
+				target: LOG_TARGET,
+				"All `ProtocolHandle`s have terminated, terminating `PeerSet`."
+			);
+			return Poll::Ready(None)
 		}
+
+		Poll::Pending
 	}
 }
 
 /// Reason for calling [`Peerset::dropped`].
+#[derive(Debug)]
 pub enum DropReason {
 	/// Substream or connection has been closed for an unknown reason.
 	Unknown,
 	/// Substream or connection has been explicitly refused by the target. In other words, the
 	/// peer doesn't actually belong to this set.
-	///
-	/// This has the side effect of calling [`PeersetHandle::remove_from_peers_set`].
 	Refused,
 }
 
+/*
```

Comment:
What is the plan with these?

Reply:
Just missed. Thanks for pointing out, I will try to adapt these tests to the new interface.